### PR TITLE
[ADD] dotnet solution migrate

### PR DESCRIPTION
**Pull request by edvilme**
Created `dotnet solution migrate solutionfile.sln` to generate a new solution file with the .slnx extension using https://github.com/microsoft/vs-solutionpersistence

Contributes to #40913 (`dotnet sln migrate`)

### CODE REVIEW

**Comment by edvilme** [comment]
> unit tests upcoming?

Hello. Yes, this is still WIP and will have a unit test that verifies if the command work. Additional testing on _how_ the conversion is done, is more related to the https://github.com/microsoft/vs-solutionpersistence project as we just consume their API

File: src/Cli/dotnet/commands/dotnet-sln/migrate/SlnMigrateCommand.cs
Comment (nagilson) [inline src/Cli/dotnet/commands/dotnet-sln/migrate/SlnMigrateCommand.cs:39]: I believe you will want to call `RunSynchronously()` on this `Task`. But I haven't used Tasks in eons, so I would check with this.

File: src/Cli/dotnet/commands/dotnet-sln/migrate/SlnMigrateCommand.cs
Comment (nagilson) [inline src/Cli/dotnet/commands/dotnet-sln/migrate/SlnMigrateCommand.cs:52]: I think you should handle the possible SolutionError which this API can throw.

File: src/Cli/dotnet/commands/dotnet-sln/migrate/SlnMigrateCommand.cs
Comment (nagilson) [inline src/Cli/dotnet/commands/dotnet-sln/migrate/SlnMigrateCommand.cs:39]: I would suggest reaching out to @/richardstanton to have them review this API call after the code has been reviewed by us to see if there's anything else they'd suggest we do. It is pretty common-place to have an internal approval from the API author. Also I have worked with them before and they are super helpful and pleasant to work with, 10/10 😉

**Review by nagilson — COMMENTED** [review]
The change looks good so far :) I've left some thoughts below. Definitely add some tests like mentioned above. Also, please keep the PR in a draft state if it's still a WIP so it's clear to reviewers.

**Comment by just-ero** [comment]
Hi, I just want to make sure: this command won't *require* the solution path to be provided, right? It will take the solution file present in the current directory if no path is specified?

After migrating (I feel like `convert` would have been a nicer name), does the original solution file get deleted? Renamed? Does it stay?

File: src/Cli/dotnet/commands/dotnet-sln/migrate/SlnMigrateCommand.cs
Comment (nagilson) [inline src/Cli/dotnet/commands/dotnet-sln/migrate/SlnMigrateCommand.cs:19]: Good work following the pattern here so that --help will be added by default :wink:

File: src/Cli/dotnet/commands/dotnet-sln/migrate/SlnMigrateCommand.cs
Comment (nagilson) [inline src/Cli/dotnet/commands/dotnet-sln/migrate/SlnMigrateCommand.cs:41]: If we handle the SolutionError to print a better message, we should handle that and make sure we don't return 0.

**Review by nagilson — COMMENTED** [review]
@just-ero raises a great point, many of our commands work if there is only one sln or project file in the directory where you don't need to specify a project or sln to take action on. Some of them even work if there are multiple and just picks one based on some order, like so:
```cs
string? potentialSln = Directory.GetFiles(arg, "*.sln", SearchOption.TopDirectoryOnly).FirstOrDefault();
```
 
 I don't think this was in the original spec but we should support that. I am very familiar with the code to do that, so if you have questions let me know :)

**Comment by nagilson** [comment]
> After migrating (I feel like `convert` would have been a nicer name), does the original solution file get deleted? Renamed? Does it stay?

@baronfel? I prefer 'migrate' personally and would keep the original solution file.

File: src/Cli/dotnet/commands/dotnet-sln/migrate/SlnMigrateCommand.cs
Comment (edvilme) [reply src/Cli/dotnet/commands/dotnet-sln/migrate/SlnMigrateCommand.cs:28]: Yes, it receives either full or relative paths and resolves to a full path

File: src/Cli/dotnet/commands/dotnet-sln/migrate/SlnMigrateCommand.cs
Comment (edvilme) [reply src/Cli/dotnet/commands/dotnet-sln/migrate/SlnMigrateCommand.cs:39]: Seems great! Will do

File: src/Cli/dotnet/commands/dotnet-sln/migrate/SlnMigrateCommandParser.cs
Comment (edvilme) [reply src/Cli/dotnet/commands/dotnet-sln/migrate/SlnMigrateCommandParser.cs:21]: I chatted with @baronfel and he told me that there were no current plans on adding more flags (at least for the time being), though I am happy to modify if needed

**Comment by edvilme** [comment]
> > After migrating (I feel like `convert` would have been a nicer name), does the original solution file get deleted? Renamed? Does it stay?
> 
> @baronfel? I prefer 'migrate' personally and would keep the original solution file, deleting the file may be unexpected and cause pain there but it could be an option to provide.

I also prefer `migrate` and definitely think we should keep the original solution file at least until the rest of the `dotnet sln`  commands are ready. Otherwise people would run the command with no way of undoing it and things could get messy (?) I would advocate towards having a `--replace` flag and additional confirmation from the user

File: src/Cli/dotnet/commands/dotnet-sln/migrate/SlnMigrateCommand.cs
Comment (edvilme) [reply src/Cli/dotnet/commands/dotnet-sln/migrate/SlnMigrateCommand.cs:19]: Thanks! It does
image

**Comment by just-ero** [comment]
> ```cs
> string? potentialSln = Directory.GetFiles(arg, "*.sln", SearchOption.TopDirectoryOnly).FirstOrDefault();
> ```

Maybe this command should panic when more than one solution file is found?

**Comment by baronfel** [comment]
I've added details to the migration section on the linked issue. We should not delete the pre-existing .sln file, users are free to do that once they have tested and verified the slnx. solves their needs. We should use our existing solution-discovery mechanisms like we do for the other `dotnet sln` commands. We should error out if there are multiple .sln files in a directory that is used and no explicit sln argument is provided.

**Comment by nagilson** [comment]
> > ```cs
> > string? potentialSln = Directory.GetFiles(arg, "*.sln", SearchOption.TopDirectoryOnly).FirstOrDefault();
> > ```
> 
> Maybe this command should panic when more than one solution file is found?

That is not how other commands behave, so I think its good to be consistent. This is the logic used by other commands.

**Comment by edvilme** [comment]
> > > ```cs
> > > string? potentialSln = Directory.GetFiles(arg, "*.sln", SearchOption.TopDirectoryOnly).FirstOrDefault();
> > > ```
> > 
> > 
> > Maybe this command should panic when more than one solution file is found?
> 
> That is not how other commands behave, so I think its good to be consistent. This is the logic used by other commands.

What I've seen so far from `SlnFileFactory` is that it outputs an error message when two or more are found.

https://github.com/dotnet/sdk/blob/b0bf7b7b4dc36fc8d7b41dac631eced050473870/src/Cli/dotnet/SlnFileFactory.cs#L68C13-L73C14

File: src/Cli/dotnet/commands/dotnet-sln/LocalizableStrings.resx
Comment (edvilme) [reply src/Cli/dotnet/commands/dotnet-sln/LocalizableStrings.resx:184]: What about 
```suggestion
    <value>.slnx solution file {0} generated.</value>
```
??

I am very open to discussing this @nagilson @baronfel

File: src/Cli/dotnet/commands/dotnet-sln/LocalizableStrings.resx
Comment (baronfel) [reply src/Cli/dotnet/commands/dotnet-sln/LocalizableStrings.resx:181]: I think we can drop the word `'solution'` from this suggestion.

File: src/Cli/dotnet/commands/dotnet-sln/LocalizableStrings.resx
Comment (baronfel) [reply src/Cli/dotnet/commands/dotnet-sln/LocalizableStrings.resx:184]: Agree with @nagilson that the .slnx part should be made more clear - I think the 'solution' part is understood in the context we're working in.

**Comment by edvilme** [comment]
Hello @dotnet/product-construction I am trying to add package `vs-solutionpersistence` but got a strong name error. Is this already signed?

**Comment by edvilme** [comment]
> to unblock, we can add this line to dotnet.csproj
> 
> https://github.com/dotnet/sdk/blob/3b653cc71df279eeb0050e427504b3d4fc40f60e/src/Containers/Microsoft.NET.Build.Containers/Microsoft.NET.Build.Containers.csproj#L20

I wouldn't choose to do that as that would not address the issue but mostly mask it, and could potentially cause issues in the future

File: src/Cli/dotnet/commands/dotnet-sln/migrate/SlnMigrateCommand.cs
Comment (nagilson) [inline src/Cli/dotnet/commands/dotnet-sln/migrate/SlnMigrateCommand.cs:37]:
```suggestion
            Task task = ConvertToSlnxAsync(slnFileFullPath, slnxFileFullPath, CancellationToken.None);
```

nit: whitespace

Comment (edvilme) [reply src/Cli/dotnet/commands/dotnet-sln/migrate/SlnMigrateCommand.cs:37]: Oh, missed it. thx! ;)

**Review by nagilson — COMMENTED** [review]
This is almost ready 🔥  I am so excited for this to go in, and I think a lot of people are :) Great work so far, the code looks clean. I have one more concern / comment, after that I think we can merge this.

File: src/Cli/dotnet/commands/dotnet-sln/migrate/SlnMigrateCommand.cs
Comment (edvilme) [reply src/Cli/dotnet/commands/dotnet-sln/migrate/SlnMigrateCommand.cs:42]: Oh I will try to wrap it under a `try` `catch`. What do you think about passing the string "The migration of file.sln to file.slnx failed due to an unrecoverable syntax error reading the file", and _also_ passing the contents of the `SolutionException.Message` as I think those could _potentially_ be relevant moving forward?

Comment (nagilson) [reply src/Cli/dotnet/commands/dotnet-sln/migrate/SlnMigrateCommand.cs:42]: I think those are good, but I would make sure that the message is defined if including it.

File: src/Cli/dotnet/commands/dotnet-sln/migrate/SlnMigrateCommand.cs
Comment (nagilson) [reply src/Cli/dotnet/commands/dotnet-sln/migrate/SlnMigrateCommand.cs]: After chatting with @rainersigwald, I believe you can call .Wait() on this task in a try catch, if it succeeds then it will continue on and you can return 0, otherwise in the catch you can catch an error and go from there.

**Review by nagilson — APPROVED** [review]
Well done, excited to see this go into the SDK 👏 

I think we can merge this, but I would still suggest to have the API owner look over the call as well as mentioned above.

**Comment by edvilme** [comment]
Hello @dotnet/source-build @richardstanton can I please get some help on the Strong Name requirement for `Microsoft.VisualStudio.SolutionPersistence`? Thx!

**Comment by richardstanton** [comment]
> Hello @dotnet/source-build @richardstanton can I please get some help on the Strong Name requirement for `Microsoft.VisualStudio.SolutionPersistence`? Thx!

My understanding was that strong naming was generally deprecated in .NET Core as it is ignored in assembly loading. (https://github.com/dotnet/runtime/blob/main/docs/project/strong-name-signing.md) So it wasn't a priority when adding to source-build. (Note, the dll should still be authenticode signed which is important)

The dll built for the NuGet package is strong named signed, partially because it is built with the .NET Framework version that requires this, but also because it is built with a very different pipeline than source-build.

It seems like suppressing the warning might be ok if everything is .NET Core.

**Comment by ellahathaway** [comment]
> > Hello @dotnet/source-build @richardstanton can I please get some help on the Strong Name requirement for `Microsoft.VisualStudio.SolutionPersistence`? Thx!
> 
> My understanding was that strong naming was generally deprecated in .NET Core as it is ignored in assembly loading. (https://github.com/dotnet/runtime/blob/main/docs/project/strong-name-signing.md) So it wasn't a priority when adding to source-build. (Note, the dll should still be authenticode signed which is important)
> 
> The dll built for the NuGet package is strong named signed, partially because it is built with the .NET Framework version that requires this, but also because it is built with a very different pipeline than source-build.
> 
> It seems like suppressing the warning might be ok if everything is .NET Core.

Agreed that the suppressing the warning is the way to go here.

**Review by nagilson — APPROVED** [review]
Looks good! I would check with @marcpopMSFT to see if we still have time to get this into 9.0.2xx and if there's any process to do so. (that was the goal, correct @baronfel?) Right now, this is targeting main, so this one you can merge regardless.

**Pull request merged**
